### PR TITLE
NKAI: rewardables

### PR DESCRIPTION
**Pull request by nullkiller**
Basic support of rewardables. Now AI visits gold warehouse.
Partially address #3584

### CODE REVIEW

**Comment by DjWarmonger** [comment]
Will look into it

**Comment by DjWarmonger** [comment]
Okay, at the very least AI picks Vases on Dunes terrain, that's easy to observe.

How about objects which cost gold (negative gold alue), such as Gazebo (Highlands), Genie's Lamp (HoTA) or Gypsy Caravan (Antagarich Burning assets:?

**Comment by nullkiller** [comment]
Negative rewards? Will take a look.

**Comment by dydzio0614** [comment]
i think complex logic would be needed to make AI judge fairly objects like shop that sells implosion spell for 30 gems

**Comment by nullkiller** [comment]
I was surprised but it works! It handles gold cost through limiters. I only fixed it not producing negative gold reward for evaluation but it actually does not matter much

**Comment by nullkiller** [comment]
[bring-genies.zip](https://github.com/user-attachments/files/15944431/bring-genies.zip)
A sample map with ancient lamps

File: AI/Nullkiller/AIUtility.cpp [hunk 1]
```diff
@@ -421,9 +421,16 @@ bool shouldVisit(const Nullkiller * ai, const CGHeroInstance * h, const CGObject
 	case Obj::BOAT:
 	case Obj::SIGN:
 		return false;
+	case Obj::SHRINE_OF_MAGIC_GESTURE:
+	case Obj::SHRINE_OF_MAGIC_INCANTATION:
+		return h->hasSpellbook();
+	case Obj::SHRINE_OF_MAGIC_THOUGHT:
+		return h->hasSpellbook() && h->getSecSkillLevel(SecondarySkill::WISDOM) >= MasteryLevel::BASIC;
+	case Obj::PYRAMID:
+		return h->hasSpellbook() && h->getSecSkillLevel(SecondarySkill::WISDOM) == MasteryLevel::EXPERT;
```
Comment (IvanSavenko) [inline AI/Nullkiller/AIUtility.cpp:430]: Kind of wish to have this somewhere inside rewardable object...

**Review by IvanSavenko — APPROVED** [review]

File: AI/Nullkiller/AIUtility.cpp (hunk 1 above)
Comment (IvanSavenko) [reply AI/Nullkiller/AIUtility.cpp:430]: I think calling `getAvailableRewards(h, Rewardable::EEventType::EVENT_FIRST_VISIT);` should tell whether it makes sense for AI to visit this object, but needs tests

Comment (nullkiller) [reply AI/Nullkiller/AIUtility.cpp:430]: Ok will take a look onto your suggestion

Comment (nullkiller) [reply AI/Nullkiller/AIUtility.cpp:430]: getAvailableRewards is protected, but exactly what I did in other places except EEventType. Have not tested it, just hero limiter. So getAvailableRewards can be used if we make it public

Comment (nullkiller) [reply AI/Nullkiller/AIUtility.cpp:430]: So can I make it public?

Comment (IvanSavenko) [reply AI/Nullkiller/AIUtility.cpp:430]: Yeah, sure. 
I usually mark stuff as private / protected to indicate that it is not used outside of the class, so generally safe to make public.

**Pull request merged**